File: data/src/main/java/org/cryptomator/data/cloud/dropbox/DropboxFolder.java
package org.cryptomator.data.cloud.dropbox;

import org.cryptomator.domain.Cloud;
import org.cryptomator.domain.CloudFolder;

class DropboxFolder implements CloudFolder, DropboxNode {

	private final DropboxFolder parent;
	private final String name;
	private final String path;

	public DropboxFolder(DropboxFolder parent, String name, String path) {
		this.parent = parent;
		this.name = name;
		this.path = path;
	}

	@Override
	public Cloud getCloud() {
		return parent.getCloud();
	}

	@Override
	public String getName() {
		return name;
	}

	@Override
	public String getPath() {
		return path;
	}

	@Override
	public DropboxFolder getParent() {
		return parent;
	}

	@Override
	public DropboxFolder withCloud(Cloud cloud) {
		return new DropboxFolder(parent.withCloud(cloud), name, path);
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/dropbox/DropboxFolder.kt
package org.cryptomator.data.cloud.dropbox

import org.cryptomator.domain.Cloud
import org.cryptomator.domain.CloudFolder

open class DropboxFolder(override val parent: DropboxFolder?, override val name: String, override val path: String) : CloudFolder, DropboxNode {

	override val cloud: Cloud?
		get() = parent?.cloud

	override fun withCloud(cloud: Cloud?): DropboxFolder? {
		return DropboxFolder(parent?.withCloud(cloud), name, path)
	}
}
